homeassistant/components/iskra/config_flow.py

```diff
@@ -0,0 +1,272 @@
+    CONF_PROTOCOL,
+    CONF_USERNAME,
+)
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import selector
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_PROTOCOL, default="Rest API"): selector.SelectSelector(
+            selector.SelectSelectorConfig(
+                options=["Rest API", "Modbus TCP"],
+                mode=selector.SelectSelectorMode.LIST,
+            ),
+        ),
+    }
+)
+
+STEP_AUTHENTICATION_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+STEP_MODBUS_TCP_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_PORT, default=10001): int,
+        vol.Required(CONF_ADDRESS, default=33): int,
+    }
+)
+
+
+async def test_rest_api_connection(host: str, authentication=None) -> BasicInfo:
+    """Check if the RestAPI requires authentication."""
+    rest_api = RestAPI(ip_address=host, authentication=authentication)
+    try:
+        basic_info = await rest_api.get_basic_info()
+    except NotAuthorised as e:
+        raise NotAuthorised from e
+    except DeviceConnectionError as e:
+        raise CannotConnect from e
+    except DeviceTimeoutError as e:
+        raise CannotConnect from e
+    except InvalidResponseCode as e:
+        raise CannotConnect from e
+    except Exception as e:
+        _LOGGER.error("Unexpected exception: %s", e)
+        raise UnknownException from e
+
+    _LOGGER.info("REST API connection successful")
+    return basic_info
+
+
+async def test_modbus_connection(host: str, port: int, address: int) -> BasicInfo:
+    """Test the Modbus connection."""
+    modbus_api = Modbus(
+        ip_address=host, protocol="tcp", port=port, modbus_address=address
+    )
+    try:
+        basic_info = await modbus_api.get_basic_info()
+    except NotAuthorised as e:
+        raise NotAuthorised from e
+    except DeviceConnectionError as e:
+        raise CannotConnect from e
+    except DeviceTimeoutError as e:
+        raise CannotConnect from e
+    except InvalidResponseCode as e:
+        raise CannotConnect from e
+    except Exception as e:
+        _LOGGER.error("Unexpected exception: %s", e)
+        raise UnknownException from e
+
+    _LOGGER.info("Modbus connection successful")
+    return basic_info
+
+
+class FlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for iskra."""
+
+    VERSION = 1
+    host: str
+    protocol: str
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self.host = user_input[CONF_HOST]
+            self.protocol = user_input[CONF_PROTOCOL]
+            if self.protocol == "Rest API":
```

Comment:
Is there a way we can detect the protocol by the host? How does the user know which one to pick?

Reply:
We can't detect the connection method, but it depends on how the user is connecting Iskra's devices. If they use a smart gateway (WiFi to RS485/IR Modbus) to connect energy meters to Ethernet, they should use the REST API. However, if they use a Modbus RTU->TCP gateway or any Iskra measuring device that supports Ethernet connectivity, they should use Modbus TCP. 

In summary:
-  REST API:  meters connected to Smart gateway 
- Modbus TCP:     meters connected to Modbus RTU->TCP gateway or Ethernet-enabled device

Reply:
But can we probe and try the other one if it fails?

Reply:
I can do something for sure but i'm not sure it would be okay. I can probe if the device responds to the REST API. If not, I can assume it is using Modbus TCP and prompt the user to enter the Modbus address and port. If that's acceptable, I can proceed with that approach. 

However, we can't be certain it should use Modbus TCP in that case, as the user might just enter the wrong IP address.

We also support UDP broadcast discovery(not standard), which could later be added as a zeroconf method. This would work for all devices, except if the user sets the smart gateway to require authentication. In such cases, the user would need to enter credentials.

Reply:
Zeroconf works with mDNS instead of UDP

Reply:
I'll have a think about the rest

Reply:
any updates?